### PR TITLE
Core: View history entry core implementation

### PR DESCRIPTION
**Pull request by amogh-jahagirdar**
Co-authored-by: John Zhuge <jzhuge@apache.org>

Separating this PR from https://github.com/apache/iceberg/pull/6559/files#diff-2a70d3056d3d0cca0da3ff4ddabc83c41c06af2296f281a5b37c5b54ead98915 for easier review from the community.

This PR contains the core implementation for view history entry and parsing/serializing logic for view history logs.

cc : @jzhuge @rdblue @jackye1995

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+package org.apache.iceberg.view;
+
+import java.util.Objects;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+
+public class BaseViewHistoryEntry implements ViewHistoryEntry {
+  private final long timestampMillis;
+  private final int versionId;
+
+  static ViewHistoryEntry of(long timestampMillis, int versionId) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java:28]: why do we need a static method if it is just delegating to the constructor?

File: core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    generator.writeNumberField(VERSION_ID, entry.versionId());
+    generator.writeEndObject();
+  }
+
+  static ViewHistoryEntry fromJson(JsonNode node) {
+    return BaseViewHistoryEntry.of(
+        JsonUtil.getLong(TIMESTAMP_MS, node), JsonUtil.getInt(VERSION_ID, node));
+  }
+
+  private ViewHistoryEntryParser() {}
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java:43]: nit: constructor should be at top?

File: core/src/test/java/org/apache/iceberg/view/ParserTestBase.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.junit.Test;
+
+public abstract class ParserTestBase<T> {
+
+  private final T entry;
+  private final String json;
+  private final JsonUtil.JsonWriter<T> writer;
+  private final JsonUtil.JsonReader<T> reader;
+
+  public ParserTestBase(
```
Comment (jackye1995) [inline core/src/test/java/org/apache/iceberg/view/ParserTestBase.java:36]: I am wondering if we really should structure the tests like this with a test base. Each object model could have different fields to be tested and this makes the tests not as flexible. Also we added all the things in the JsonUtil which I am not sure if it is useful outside this context

File: core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java (hunk 1 above)
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java:28]: It's not needed, just a part of the original CR. We can just make the constructor public and use that

Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java:28]: I see. I can see the constructor being potentially protected, we can keep it like that for now, but the static method seems unnecessary

File: core/src/test/java/org/apache/iceberg/view/ParserTestBase.java (hunk 3 above)
Comment (amogh-jahagirdar) [reply core/src/test/java/org/apache/iceberg/view/ParserTestBase.java:36]: Agreed, I think what we can do is just follow the existing pattern for testing table metadata and just enumerate different toJson/fromJson cases. While the benefit of having a parent class is to reduce some duplication, for the parsing code it seems better just to have more explicit tests (for example like we do with ref parser)

**Comment by amogh-jahagirdar** [comment]
@jackye1995 I updated the PR so it should be a bit simpler now , I agree at this point we don't need the abstractions of base test classes  . Thanks for the review! @jzhuge @rdblue let me know your thoughts on this!

**Review by jackye1995 — APPROVED** [review]
looks good to me!

**Comment by jackye1995** [comment]
Also ping a few people in #4925 to as we are starting to get all the implementations in

@rdblue @stevenzwu @danielcweeks @wmoustafa @nastra

File: core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  static ViewHistoryEntry fromJson(String json) {
+    Preconditions.checkArgument(
+        json != null && !json.isEmpty(),
+        "Cannot parse view history entry from invalid JSON: %s",
+        json);
+    return JsonUtil.parse(json, ViewHistoryEntryParser::fromJson);
+  }
+
+  static ViewHistoryEntry fromJson(JsonNode node) {
+    return new BaseViewHistoryEntry(
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java:54]: I think this requires 2 precondition checks here:

```
Preconditions.checkArgument(null != json, "Cannot parse view history entry from null object");
Preconditions.checkArgument(json.isObject(), "Cannot parse view history entry from non-object: %s", json);
```

and then the `fromJson(String json)` method could be changed to 

```
static ViewHistoryEntry fromJson(String json) {
    return JsonUtil.parse(json, ViewHistoryEntryParser::fromJson);
  }
```

File: core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  static final String VERSION_ID = "version-id";
+  static final String TIMESTAMP_MS = "timestamp-ms";
+
+  static String toJson(ViewHistoryEntry entry) {
+    return JsonUtil.generate(gen -> toJson(entry, gen), false);
+  }
+
+  static void toJson(ViewHistoryEntry entry, JsonGenerator generator) throws IOException {
+    generator.writeStartObject();
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java:39]: this should have a precondition check: `Preconditions.checkArgument(null != entry, "Invalid view history entry: null");

File: core/src/test/java/org/apache/iceberg/view/TestViewHistoryEntryParser.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * under the License.
+ */
+package org.apache.iceberg.view;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestViewHistoryEntryParser {
+
+  @Test
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/view/TestViewHistoryEntryParser.java:26]: this needs are few more tests around null and empty json and such. Some examples can be found in [`TestCounterResultParser`](https://github.com/apache/iceberg/blob/master/core/src/test/java/org/apache/iceberg/metrics/TestCounterResultParser.java#L28-L48)

**Review by nastra — COMMENTED** [review]
overall LGTM but needs a few small updates

File: core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java (hunk 4 above)
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java]: Separated the precondition checks.

File: core/src/test/java/org/apache/iceberg/view/TestViewHistoryEntryParser.java (hunk 6 above)
Comment (amogh-jahagirdar) [reply core/src/test/java/org/apache/iceberg/view/TestViewHistoryEntryParser.java:28]: Good point, I've added more tests around null,empty, missing fields

File: core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java (hunk 5 above)
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/view/ViewHistoryEntryParser.java:40]: Sure, updated

File: core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.view;
+
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface BaseViewHistoryEntry extends ViewHistoryEntry {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java:24]: I think we can drop `Base` since this isn't a base class. We use that when we expect something to extend the class. But this is an interface that will result in an `ImmutableViewHistoryEntry`.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good other than the name of the history entry interface. Thanks, @amogh-jahagirdar!

File: core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java (hunk 7 above)
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java:24]: Sure let me see about a better name, the underlying interface is already ViewHistoryEntry. I think technically we could just need to add the immutable to ViewHistoryEntry but maybe its better to have an explicit core implementation which uses the annotation.

Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java]: maybe a dumb question, why don’t we just add immutable annotation on the interface in that case?

```
@Value.Immutable
public interface ViewLogEntry extends ViewHistoryEntry {
  long timestampMillis();

  int versionId();
}
```

```
public interface ViewHistoryEntry {
  /** Return the timestamp in milliseconds of the change */
  long timestampMillis();

  /** Return ID of the new current version */
  int versionId();
}
```

Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/view/BaseViewHistoryEntry.java]: yeah nvm on the classpath issue I mentioned, I confused myself. We can just add to the base interface and eliminate the duplicate interface

**Review by jackye1995 — APPROVED** [review]

**Comment by jackye1995** [comment]
Since the immutable is addressed and we have enough approvals, I will go ahead to merge the PR, thanks for the review @rdblue and @nastra !

**Comment by jzhuge** [comment]
LGTM!
